### PR TITLE
Lezhin : Login & pictures fix

### LINKED ISSUE #2950: [Lezhin Korean] Connector not working
**Did you read [the troubleshooting guide](https://hakuneko.download/docs/troubleshoot/)**
Yes

**Is the website of the connector working properly / are you able to see the manga within your browser**
Yes

**Describe the bug**
The whole site works. Just specifically for this one manga, paid chapters don't work.

**To Reproduce**
1. Lezhin Korean
2. 딸도둑이 될래!
3. chapter 24
Screen Shot 2021-01-18 at 3 07 38 PM

4. Copy any relevant error message that appears in a popup or (advanced) within the console (press F12)

**Screenshots**
If complex to describe, add screenshots to help explain your problem.

**HakuNeko (please complete the following information):**
 - Version [e.g. 6.17]

### LINKED ISSUE #4433: [Lezhin US] Connector not working
**Did you read [the troubleshooting guide](https://hakuneko.download/docs/troubleshoot/)**
Yes

**Is the website of the connector working properly / are you able to see the manga within your browser**
Yes

**Describe the bug**
I can't access chapters beyond episodes 1 and special episodes. Only the free ones are viewable, is there any way for me to use my purchased chapters in hakuneko?

**To Reproduce**
1. Lezhin US
2. Mad Place (if connector update is working)
3. All of them minus the free ones (if manga update is working)
4. Copy any relevant error message that appears in a popup or (advanced) within the console (press F12)

**Screenshots**
image

**HakuNeko (please complete the following information):**
 - Version [e.g. 6.17]

### LINKED ISSUE #5044: [Lezhin] connector not working
Hi

https://www.lezhin.com/ko/comic/put_it_in_the_sun

I am trying to download this manga and I bought it on lezhin. I used to be able to download lezhin chapters a month ago, but now I can't.

I logged into lezhin inside hakuneko program, so I did login to lezhin website process inside hakuneko.

### PR DESCRIPTION
**Pull request by MikeZeDev**
* Properly set Auth header in case of user logged
* Handle user logout
* Sanitize user password (edge case)
* Download pictures urls need Policy Signature and Key-Pair-Id
* Fixes https://github.com/manga-download/hakuneko/issues/5044, Fixes https://github.com/manga-download/hakuneko/issues/4433, fixes https://github.com/manga-download/hakuneko/issues/2950
* This need more testing for bought / unlocked chapters
* Tested working with EN JP KO website. We dont support ES and ES use another template, similar to.. Delitoon :D

### CODE REVIEW

File: src/web/mjs/connectors/templates/Lezhin.mjs [hunk 1]
```diff
@@ -1,19 +1,17 @@
...
-
     constructor() {
         super();
         super.id = undefined;
         super.label = undefined;
         this.tags = [];
         this.url = undefined;
-        this.apiURL = 'https://www.lezhin.com/api/v2'; // https://api.lezhin.com/v2
+        this.apiURL = 'https://www.lezhinus.com';
+        // /lz-api/v2/
```
Comment (27136773) [inline src/web/mjs/connectors/templates/Lezhin.mjs]: Remove comments of minor value

File: src/web/mjs/connectors/templates/Lezhin.mjs [hunk 2]
```diff
@@ -35,31 +33,57 @@ export default class Lezhin extends Connector {
...
+            if (!data.token) {
+                this.requestOptions.headers.delete('Authorization');
+                this.userID = '';
+            }
+        }
+
         if(this.userID || !this.config.username.value || !this.config.password.value) {
             return;
         }
+        const password = this.config.password.value.replace("'", "\\'"); //escape the password, because if it contains a single quote the script will fail
```
Comment (27136773) [inline src/web/mjs/connectors/templates/Lezhin.mjs:61]: Nice finding

File: src/web/mjs/connectors/templates/Lezhin.mjs [hunk 3]
```diff
@@ -35,31 +33,57 @@ export default class Lezhin extends Connector {
...
-                        type: 'POST',
-                        url: form.prop('action'),
-                        data: form.serialize(),
-                        success: resolve,
-                        error: reject
-                    });
-                } catch(error) {
-                    reject(error);
+        new Promise((resolve, reject) => {
+            try {
```
Comment (27136773) [inline src/web/mjs/connectors/templates/Lezhin.mjs]: The `try/catch` block might be redundant as a promise will automatically `reject` when a non-caught error occurs (in synchronous code)

File: src/web/mjs/connectors/templates/Lezhin.mjs [hunk 4]
```diff
@@ -68,165 +92,153 @@ export default class Lezhin extends Connector {
...
+    async _getMangas() {
+        let mangaList = [];
+        for (let page = 0, run = true; run; page++) {
+            let mangas = await this._getMangasFromPage(page);
+            mangas.length > 0 ? mangaList.push(...mangas) : run = false;
+        }
+        return mangaList;
     }
+    async _getMangasFromPage(page) {
+        //https://www.lezhinus.com/lz-api/v2/contents?menu=general&limit=36&offset=0&order=popular
```
Comment (27136773) [inline src/web/mjs/connectors/templates/Lezhin.mjs]: The comment may be removed, everything can easily be inferred from the code below

File: src/web/mjs/connectors/templates/Lezhin.mjs [hunk 5]
```diff
@@ -68,165 +92,153 @@ export default class Lezhin extends Connector {
...
+
+        const episode = data.data.extra.episode;
+        let purchased = episode.coin == 0;
+        //purchased = purchased || (episode.freedAt && episode.freedAt < Date.now());
+        const subscribed = data.data.extra.subscribed;
+        const updatedAt = episode.updatedAt;
+
+        let pictures = [];
+
+        for (let i = 0; i < episode.scrollsInfo.length; i++) {
```
Comment (27136773) [inline src/web/mjs/connectors/templates/Lezhin.mjs]: Get pictures is supposed to be fast, but now it makes a request for each image which may take a long time.
I recommend to leverage the `_handleConnectorURI` mechanism for requesting the image data and make these requests there.

File: src/web/mjs/connectors/templates/Lezhin.mjs (hunk 2 above)
Comment (MikeZeDev) [reply src/web/mjs/connectors/templates/Lezhin.mjs:61]: Pure luck but yeah

File: src/web/mjs/connectors/templates/Lezhin.mjs (hunk 3 above)
Comment (MikeZeDev) [reply src/web/mjs/connectors/templates/Lezhin.mjs]: I just mimicked other promisse, but that makes sense.

**Review by 27136773 — APPROVED** [review]

**Pull request merged**